### PR TITLE
[GEOS-10542] Use headers for proxy URL on workspace level

### PR DESCRIPTION
**Pull request by ahmedababnehgeo**
[GEOS-10542](https://osgeo-org.atlassian.net/browse/GEOS-10542)

This PR aims at adding a new checkbox on the workspace level to allow adding headers for the proxy URL. This will allow end users to override the disabled checkbox "Use headers for Proxy URL" on the global settings page, by checking this option per workspace.
  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [x] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by ahmedababnehgeo** [comment]
@fernandor777 PR ready for review

File: src/main/src/main/java/org/geoserver/ows/ProxifyingURLMangler.java
Comment (fernandor777) [inline src/main/src/main/java/org/geoserver/ows/ProxifyingURLMangler.java:106]: Please add some FINE level logging to know whats happening and the result of this logic conditions execution.

**Review by fernandor777 — COMMENTED** [review]
Overall looks good, just some quick request for logging lines.

File: src/main/src/main/java/org/geoserver/ows/ProxifyingURLMangler.java
Comment (groldan) [reply src/main/src/main/java/org/geoserver/ows/ProxifyingURLMangler.java:106]: While it seems ok to set the flag in `SettingsInfo`, there's still an impedance mismatch with `GeoServer.getGlobal().getSettings():SeettingsInfo`, on which these would be of no use.
If `isUseHeadersProxyURL` is **moved** instead to `SettingsInfo`, everything would be consistent and this decision making unnecessary, leaving just `doMangleHeaders = this.geoServer.getSettings().isUseHeadersProxyURL()`, and letting `geoServer.getSettings()` decide whether to return the workspace or the global settings.

That of course implies deprecating `GeoServerInfo.is/setUseHeadersProxyURL` and defer to its internal `SettingsInfo` (e.g. `@Deprecated public Boolean isUseHeadersProxyURL() { return getSettings().isUseHeadersProxyURL();}`)

**Review by groldan — CHANGES_REQUESTED** [review]

File: src/main/src/main/java/org/geoserver/ows/ProxifyingURLMangler.java
Comment (groldan) [reply src/main/src/main/java/org/geoserver/ows/ProxifyingURLMangler.java:106]: @aaime thoughts?

Comment (groldan) [reply src/main/src/main/java/org/geoserver/ows/ProxifyingURLMangler.java:106]: hmm thinking further, still the workspace settings may not have the value set at all and you'll still need to manually select the global settings value. For which you would need to define it as a `Boolean` instead of `boolean` and allow it to be `null`. Maybe add an `Optional<Boolean> useUseHeadersProxyURL(){return Optional.ofNullable(this.useHeadersProxyURL);}`

**Comment by aaime** [comment]
I agree that ``useHeadersForProxyURL`` should not have been placed inside GeoServerInfo, but inside SettingsInfo, where the proxy base URL also is. Deprecating and delegating sounds like a good plan, but there is a hole: upgrading users.

The current setting has been around for 4 years now, it must keep on working as people upgrade from older data directories.
It is also going to break whoever is using the REST API, since XStream goes directly onto the fields, skipping the getters/setters.

One possible avenue is to write a custom XStream converter, that would migrate the value from the GlobalInfoImpl, to its settings, on deserialization. Could be a subclass of AbstractReflectionConverter in XStreamPersister, and dedicated to the serialization of GeoServerInfoImpl.

**Comment by aaime** [comment]
Looks good to me, thanks!

**Comment by jodygarnett** [comment]
I cam to this pull request because this is listed as a new feature for GeoServer 2.21.1 ... but I do not see docs?

Please do not merge a pull request with a user interface change with out docs.

**Pull request merged**
